projects/telemetry/src/nmea.c

```diff
@@ -0,0 +1,264 @@
+  }
+  if (to_check[0] != '$') {
+    return status_msg(STATUS_CODE_UNKNOWN, "First character of a NMEA sentence should be a $\n");
+  }
+  if (to_check[1] != 'G' || to_check[2] != 'P') {
+    return status_msg(STATUS_CODE_UNKNOWN, "NMEA sentence should begin with GP\n");
+  }
+  if (!nmea_checksum_validate((char *)to_check, len)) {
+    return status_msg(STATUS_CODE_UNKNOWN, "Invalid checksum for NMEA message\n");
+  }
+  return STATUS_CODE_OK;
+}
+
+// Returns the NMEA sentence type
+StatusCode nmea_sentence_type(const char *rx_arr, NmeaMessageID *result) {
+  if (result == NULL || rx_arr == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Cannot pass NULL pointers to this function\n");
+  }
+
+  size_t len = strlen(rx_arr);
+
+  // Array index 3 should be 0, so that it is a null terminated string
+  char message_id[4] = { 0 };
+  for (uint16_t i = 3; i < len && i < 6; i++) {
+    if (rx_arr[i] == ',') {
+      break;
+    }
+    message_id[i - 3] = rx_arr[i];
+  }
+
+  // Making sure array index 3 is \n
+  message_id[3] = '\0';
+
+  // Parsing which type of NMEA message this is
+  if (strcmp(message_id, "GGA") == 0) {
+    *result = NMEA_GGA;
+    return STATUS_CODE_OK;
+  } else if (strcmp(message_id, "GLL") == 0) {
+    *result = NMEA_GLL;
+    return STATUS_CODE_OK;
+  } else if (strcmp(message_id, "GSA") == 0) {
+    *result = NMEA_GSA;
+    return STATUS_CODE_OK;
+  } else if (strcmp(message_id, "GSV") == 0) {
+    *result = NMEA_GSV;
+    return STATUS_CODE_OK;
+  } else if (strcmp(message_id, "RMC") == 0) {
+    *result = NMEA_RMC;
+    return STATUS_CODE_OK;
+  } else if (strcmp(message_id, "VTG") == 0) {
+    *result = NMEA_VTG;
+    return STATUS_CODE_OK;
+  } else {
+    LOG_DEBUG("Unknown message type: %c%c%c", message_id[0], message_id[1], message_id[2]);
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unknown NMEA message type\n");
+  }
+}
+
+StatusCode nmea_get_gga_sentence(const char *rx_arr, nmea_gga_sentence *result) {
+  if (result == NULL || rx_arr == NULL) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Cannot pass NULL pointers to this function\n");
+  }
+
+  size_t len = strlen(rx_arr);
+
+  status_ok_or_return(nmea_valid(rx_arr));
+
+  // m_id will keep track of which sentence type we are currently operating on
+  NmeaMessageID m_id = 0;
+
+  status_ok_or_return(nmea_sentence_type(rx_arr, &m_id));
+
+  result->message_id = m_id;
+
+  // Parses NMEA message
+
+  if (m_id == NMEA_GGA) {
+    // Example message:
+    // $GPGGA,053740.000,2503.6319,N,12136.0099,E,1,08,1.1,63.8,M,15.2,M,,0000*64
+
+    char *rx_arr_copy = strdup(rx_arr);
+
+    // strsep modifies original pointer, so we have to keep a copy
+    char *rx_arr_copy_pointer_backup = rx_arr_copy;
+
+    // Get rid of $GPGGA
+    char *token = strsep(&rx_arr_copy, ",");
+
+    token = strsep(&rx_arr_copy, ",");
+    if (token != NULL) {
+      sscanf(token, "%2" SCNd8 "%2" SCNd8 "%2" SCNd8 ".%3" SCNd16, &result->time.hh,
```

Comment:
If you're using sscanf, why not just populate the whole thing using it?

Reply:
Doesn't seem like there is a straight forward approach to handling two consecutive delimiters. One solution I thought of was replacing `,,` with `,0,` but this doesn't work for all fields as some are floats, so they would need `,0.0,`. Then also some fields are chars, so this approach would not work for those either. Eventually it would come out to a similar complexity so I just did it the intuitive way